_posts/2018-05-04-design-of-task-graph-generation.md

```diff
@@ -0,0 +1,253 @@
+[`.cron.yml`](https://dxr.mozilla.org/mozilla-central/source/.cron.yml) in the
+root of the source tree. It then creates a decision task for each matching
+entry, with a custom task-graph filter configuration to select only the desired
+tasks.
+
+## Actions
+
+For the most part, the task-graph for a push (or cron task) is defined in
+advance. But developers and sheriffs often need to modify a task-graph after it
+is created, for example to retrigger a task or run a test that was accidentally
+omitted from a try push. Taskcluster defines a generic notion of an
+["action"](https://docs.taskcluster.net/manual/using/actions/spec) for just
+this purpose: acting on an existing task-graph.
+
+Briefly, the decision task publishes a description of the actions that are
+available for the tasks in the task-graph. Services like Treeherder and the
+Taskcluster task inspector then use that description to connect user-interface
+elements to those actions. When an action is executed, the user interface
+creates a new task called an *action task* that performs the desired action.
+
+Action tasks are similar to decision and cron tasks: they clone the desired
+revision of the source code, then run a mach command to do whatever the user
+has requested.
+
+## Multiple Projects
+
+The task-graph generation code rides the trees, with occasional uplifts, just
+like the rest of the Firefox codebase. That means that the same code must work
+correctly for all branches; we do not have a different implementation for the
+mozilla-beta branch, for example.
+
+While it might seem like, to run a new task on mozilla-central, you would just
+land a patch adding that task on mozilla-central, it's not that simple: without
+adjusting the filtering, that task would eventually be merged to all other
+projects and execute everywhere.
+
+This also makes testing tricky: since the task-graph generation is different
+for every project, it's possible to land code which works fine in try and
+inbound, but fails on mozilla-central. It is easy to test task-graph generation
+against specific situations (all inputs to the process are encapsulated in a
+`parameters.yml` file easily downloaded from a decision task). The artistry is
+in figuring out which situations to test.
+
+## Try Pushes
+
+Pushes to try trigger decision tasks just like any other project, but the
+filtering process is a little more complex.
+
+If the push comes with legacy try syntax (`-b do -p win64,linux64 -u
+all[linux64-qr,windows10-64-qr] -t all[linux64-qr,windows10-64-qr]` - clear as
+mud, right?), we do our best to emulate the behavior of the Buildbot try parser
+in filtering out tasks that were not requested. The legacy syntax is deeply
+dependent on some Buildbot-specific features, and does not cover new
+functionality like optimization, so there are lots of edge cases where it
+behaves strangely or does not work at all.
+
+The better alternative is try-select, where the push contains a
```

Comment:
Maybe write `./mach try fuzzy` I'm sure there's a lot of readers who is unaware of this feature.